helm/config.go

```diff
@@ -0,0 +1,401 @@
+    tcp.Add("name", NewScalar("\"{{ $key }}-tcp\""))
+    tcp.Add("containerPort", NewScalar("$key"))
+    tcp.Add("protocol", NewScalar("TCP"))
+
+    ports := NewList(Comment("List of TCP ports"))
+    ports.Add(tcp)
+    root.Add("ports", ports)
+
+  Generates this document:
+
+    # List of TCP ports
+    ports:
+    {{- range $key, $value := .Values.tcp }}
+    - name: "{{ $key }}-tcp"
+      containerPort: $key
+      protocol: TCP
+    {{- end }}
+
+*/
+package helm
+
+import (
+	"fmt"
+	"io"
+	"sort"
+	"strings"
+)
+
+// sharedFields provides the shared metadata (comments & block actions) for all
+// node types.
+type sharedFields struct {
+	block   string
+	comment string
+}
+
+// NodeModifier functions can be used to set values of shared fields in a node.
+type NodeModifier func(*sharedFields)
+
+// Block returns a modifier function to set the block action of a node.
+func Block(block string) func(*sharedFields) {
+	return func(shared *sharedFields) { shared.block = block }
+}
+
+// Comment returns a modifier function to set the comment of a node.
+func Comment(comment string) func(*sharedFields) {
+	return func(shared *sharedFields) { shared.comment = comment }
+}
+
+// Set applies NodeModifier functions to the embedded sharedFields struct.
+func (shared *sharedFields) Set(modifiers ...NodeModifier) {
+	for _, modifier := range modifiers {
+		modifier(shared)
+	}
+}
+
+// Block is a shared getter function for all node types.
+func (shared sharedFields) Block() string {
+	return shared.block
+}
+
+// Comment is a shared getter function for all node types.
+func (shared sharedFields) Comment() string {
+	return shared.comment
+}
+
+// Node is the interface implemented by all config node types.
+type Node interface {
+	// Every node will embed a sharedFields struct and inherit these methods:
+	Block() string
+	Comment() string
+	Set(...NodeModifier)
+
+	// The write() method implements the part of Encoder.writeNode() that
+	// needs to access the fields specific to each node type. prefix will be
+	// the indented label, either "name:" or "-", depending on whether the
+	// node is part of an object or a list.
+	write(enc *Encoder, prefix string)
+}
+
+// Scalar represents a scalar value inside a list or object.
+type Scalar struct {
+	sharedFields
+	value string
+}
+
+// NewScalar creates a scalar node and initializes shared fields.
+func NewScalar(value string, modifiers ...NodeModifier) *Scalar {
+	scalar := &Scalar{value: value}
+	scalar.Set(modifiers...)
+	return scalar
+}
+
+func (scalar Scalar) write(enc *Encoder, prefix string) {
+	if !strings.ContainsRune(scalar.value, '\n') {
+		fmt.Fprintln(enc, prefix+" "+scalar.value)
+		return
+	}
+
+	// Scalars including newlines will be written using the "literal" YAML format.
+	fmt.Fprintln(enc, prefix+" |-")
+	// Calculate proper indentation for data lines.
+	if strings.HasSuffix(prefix, ":") {
+		prefix = strings.Repeat(" ", strings.LastIndex(prefix, " ")+1+enc.indent)
```

Comment:
Actually, what happens if the key contains a space?
```yaml
---
foo:
  "bar baz": 3
```

Reply:
This is currently not supported. We could fix this (the quotes are not even required here).

But we would need to add quotes if we want to support keys containing `:` characters.

Not sure this is worth it, at least not as long as this is just a package inside fissile, and not a standalone thing.

Maybe file an issue for this once this PR is merged, for future consideration?

Reply:
I'd be happy with just having the limitations documented.